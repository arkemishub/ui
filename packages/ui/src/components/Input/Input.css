/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.input__container {
    @apply z-10 w-full px-2 flex items-center min-h-[40px] min-w-[120px] border border-neutral rounded-theme bg-background text-background-contrast focus:ring-0 focus:outline-offset-0 focus:outline-primary disabled:opacity-50 disabled:cursor-not-allowed
}

.input__container:has(> .input:focus){
    @apply outline-offset-0 border border-primary outline outline-2 outline-primary/30 disabled:cursor-not-allowed
}

.input__container:has(> .input:disabled){
    @apply opacity-50 cursor-not-allowed
}

.input {
    @apply bg-transparent min-h-[40px] min-w-[120px] w-full !px-2 !outline-0 !border-0
}

.label {
    @apply mb-1 block text-sm
}

.input--error {
    @apply outline outline-error focus:outline-error
}

.input__helperText {
    @apply mt-2 text-sm opacity-50
}

.input__helperText--error {
    @apply opacity-100 text-error
}

.input__startAdornment {
    @apply cursor-pointer ml-1
}

.input__endAdornment {
    @apply cursor-pointer mr-1
}

.input__prefixAdornment {
    @apply flex items-center px-2 border border-neutral rounded-theme w-fit rounded-tr-none rounded-br-none border-r-0
}

.input__prefixAdornment .select__options {
    @apply !w-fit
}

.input__suffixAdornment {
    @apply flex items-center px-2 border border-neutral rounded-theme w-fit rounded-tl-none rounded-bl-none border-l-0
}

.input__suffixAdornment .select__options {
    @apply !w-fit
}