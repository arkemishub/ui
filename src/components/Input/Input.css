/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.input__container {
    @apply relative inline-block
}

.input__startAdornment {
    @apply absolute left-0 top-0 h-full flex items-center p-2
}

.input__endAdornment {
    @apply absolute right-0 top-0 h-full flex items-center p-2
}

.input {
    @apply min-h-[40px] min-w-[120px] border border-neutral rounded-theme bg-background text-background-contrast px-4 focus:ring-0 focus:outline-offset-0 focus:outline-primary disabled:opacity-50 disabled:cursor-not-allowed
}

.label {
    @apply mb-1 block text-sm
}

.input--error {
    @apply outline outline-error focus:outline-error
}

.input__helperText {
    @apply mt-2 text-sm opacity-50
}

.input__helperText--error {
    @apply opacity-100 text-error
}
