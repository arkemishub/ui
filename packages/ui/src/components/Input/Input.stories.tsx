/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Input } from "./index";
import React, { ChangeEventHandler, ReactNode, useState } from "react";
import { IInputProps } from "./Input.types";
import { useArgs } from "@storybook/client-api";
import {
  CopyIcon,
  CreditCardIcon,
  EyeIcon,
  EyeSlashIcon,
  InfoIcon,
  ItaIcon,
  PhoneIcon,
  UsaIcon,
  UserIcon,
} from "../../storiesUtils/Icons";
import { Select } from "../Select";

export default {
  title: "Input",
  component: Input,
  argTypes: {
    value: {
      defaultValue: "",
    },
    label: {
      defaultValue: "Label",
    },
    name: {
      defaultValue: "test",
    },
  },
};

export const Default = (args: IInputProps) => {
  const [{ value }, updateArgs] = useArgs();

  const handleChange: ChangeEventHandler<HTMLInputElement> = (e) =>
    updateArgs({ value: e.target.value });

  return (
    <>
      <Input {...args} value={value} onChange={handleChange} />
      <Input {...args} disabled />
    </>
  );
};

export const Adornments = (args: IInputProps) => {
  const [showPassword, setShowPassword] = useState<boolean>(true);
  const [{ value }, updateArgs] = useArgs();

  const handleChange: ChangeEventHandler<HTMLInputElement> = (e) =>
    updateArgs({ value: e.target.value });

  return (
    <div className="flex flex-col gap-2">
      <Input
        label="Username"
        placeholder="Add your username"
        startAdornment={<UserIcon />}
      />
      <Input
        label="Password"
        placeholder="**********"
        value={value}
        type={showPassword ? "text" : "password"}
        onChange={handleChange}
        endAdornment={
          <span
            role="presentation"
            onClick={() => setShowPassword(!showPassword)}
          >
            {showPassword ? <EyeSlashIcon /> : <EyeIcon />}
          </span>
        }
      />
      <Input
        label="Username"
        placeholder="Add your username"
        startAdornment={<UserIcon />}
        endAdornment={<InfoIcon />}
      />
      <Input
        label="Website"
        placeholder="Add your credit website"
        prefixAdornment={<div>https://</div>}
        suffixAdornment={<div>Copy</div>}
      />
      <Input
        hasError
        label="Credit Card"
        placeholder="Add your credit card name"
        helperText="Insert your credit card"
        endAdornment={<CreditCardIcon />}
      />
      <Input
        label="Unit"
        placeholder="Add your measure unit"
        suffixAdornment={"px"}
      />
      <Input
        label="Full optional"
        placeholder="Add your text"
        helperText="This is an amazing input"
        startAdornment={<EyeIcon />}
        endAdornment={<InfoIcon />}
        prefixAdornment={<div>https://</div>}
        suffixAdornment={
          <div className="flex items-center gap-1">
            <CopyIcon />
            <p>Copy</p>
          </div>
        }
      />
    </div>
  );
};

export const Complex = (args: IInputProps) => {
  const [currency, setCurrency] = useState<{ value: string; label: string }>();
  const [phonePrefix, setPhonePrefix] = useState<{
    value: string;
    label: string;
    icon(): ReactNode;
  }>();

  const currencyValues = [
    { label: "EUR", value: "euro" },
    { label: "USD", value: "usd" },
  ];

  const phoneValues = [
    {
      icon: () => <ItaIcon className="w-4 h-4" />,
      label: "Italia",
      value: "it",
    },
    {
      icon: () => <UsaIcon className="w-4 h-4" />,
      label: "United States",
      value: "en",
    },
  ];

  return (
    <div className="flex flex-col gap-2">
      <Input
        label="Currency"
        placeholder="Insert value"
        helperText="This is an amazing currency input"
        endAdornment={<InfoIcon />}
        suffixAdornment={
          <Select
            values={currencyValues}
            value={currency}
            onChange={(val) => setCurrency(val)}
            renderLabel={(val) => val.label}
            className="border-none min-w-[inherit]"
          />
        }
      />
      <Input
        label="Phone"
        placeholder="Add your phone number"
        startAdornment={<PhoneIcon className="w-5 h-5" />}
        prefixAdornment={
          <Select
            values={phoneValues}
            value={phonePrefix}
            onChange={(val) => setPhonePrefix(val)}
            renderLabel={(val) => val.icon()}
            renderOption={(val) => (
              <div className="flex items-center gap-2 whitespace-nowrap">
                {val.icon()}
                {val.label}
              </div>
            )}
            className="p-0 border-none min-w-[inherit]"
          />
        }
      />
    </div>
  );
};
