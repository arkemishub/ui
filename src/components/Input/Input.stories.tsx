/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Input } from "./index";
import React, { ChangeEventHandler } from "react";
import { IInputProps } from "./Input.types";
import { useArgs } from "@storybook/client-api";

export default {
  title: "Input",
  component: Input,
  argTypes: {
    value: {
      defaultValue: "",
    },
    label: {
      defaultValue: "Label",
    },
    name: {
      defaultValue: "test",
    },
  },
};

const Icon = () => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    fill="none"
    viewBox="0 0 24 24"
    strokeWidth="1.5"
    stroke="currentColor"
    className="w-6 h-6"
  >
    <path
      strokeLinecap="round"
      strokeLinejoin="round"
      d="M4.26 10.147a60.436 60.436 0 00-.491 6.347A48.627 48.627 0 0112 20.904a48.627 48.627 0 018.232-4.41 60.46 60.46 0 00-.491-6.347m-15.482 0a50.57 50.57 0 00-2.658-.813A59.905 59.905 0 0112 3.493a59.902 59.902 0 0110.399 5.84c-.896.248-1.783.52-2.658.814m-15.482 0A50.697 50.697 0 0112 13.489a50.702 50.702 0 017.74-3.342M6.75 15a.75.75 0 100-1.5.75.75 0 000 1.5zm0 0v-3.675A55.378 55.378 0 0112 8.443m-7.007 11.55A5.981 5.981 0 006.75 15.75v-1.5"
    />
  </svg>
);

export const Default = (args: IInputProps) => {
  const [{ value }, updateArgs] = useArgs();

  const handleChange: ChangeEventHandler<HTMLInputElement> = (e) =>
    updateArgs({ value: e.target.value });

  return <Input {...args} value={value} onChange={handleChange} />;
};

export const StartEndAdornments = (args: IInputProps) => {
  const [{ value }, updateArgs] = useArgs();

  const handleChange: ChangeEventHandler<HTMLInputElement> = (e) =>
    updateArgs({ value: e.target.value });

  return (
    <Input
      {...args}
      value={value}
      startAdornment={<Icon />}
      endAdornment={<Icon />}
      onChange={handleChange}
    />
  );
};
