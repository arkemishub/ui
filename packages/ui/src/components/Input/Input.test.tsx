/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Input from "./Input";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("Input", () => {
  test("should match snapshot", () => {
    const { asFragment } = render(
      <Input value="Value" type="text" onChange={() => null} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render input", () => {
    const { getByTestId } = render(
      <Input value="Value" type="text" onChange={() => null} />
    );
    expect(getByTestId("arke-container")).toBeInTheDocument();
  });

  test("should call onChange when input changes", async () => {
    const onChange = jest.fn();
    const { getByTestId } = render(
      <Input value="Value" type="text" onChange={onChange} />
    );
    await userEvent.type(getByTestId("arke-input"), "Test");
    expect(onChange).toHaveBeenCalled();
  });

  test("should render label", () => {
    const { getByText } = render(
      <Input value="Value" type="text" onChange={() => null} label="Label" />
    );
    expect(getByText("Label")).toBeInTheDocument();
  });

  test("should render helper text", () => {
    const { getByText } = render(
      <Input
        value="Value"
        type="text"
        onChange={() => null}
        helperText="Helper text"
      />
    );
    expect(getByText("Helper text")).toBeInTheDocument();
  });

  test("should render error input when hasError is true", () => {
    const { getByTestId } = render(
      <Input value="Value" type="text" onChange={() => null} hasError />
    );
    expect(getByTestId("arke-container")).toHaveClass("input--error");
  });

  test("should render full width input when fullWidth is true", () => {
    const { getByTestId } = render(
      <Input value="Value" type="text" onChange={() => null} fullWidth />
    );

    expect(getByTestId("arke-container")).toHaveClass("w-full");
  });

  test("should render helperText with error class when hasError is true", () => {
    const { getByText } = render(
      <Input
        value="Value"
        type="text"
        onChange={() => null}
        helperText="Helper text"
        hasError
      />
    );
    expect(getByText("Helper text")).toHaveClass("input__helperText--error");
  });

  test("should add custom class when className is passed", () => {
    const { getByTestId } = render(
      <Input
        value="Value"
        type="text"
        onChange={() => null}
        className="custom-class"
      />
    );
    expect(getByTestId("arke-container")).toHaveClass("custom-class");
  });

  test("should contain start/end Adornment when props is passed", () => {
    const { container } = render(
      <Input
        value="Value"
        type="text"
        onChange={() => null}
        startAdornment={"start adornment"}
        endAdornment={"end adornment"}
      />
    );
    expect(
      container.getElementsByClassName("input__startAdornment").length
    ).toBe(1);
    expect(container.getElementsByClassName("input__endAdornment").length).toBe(
      1
    );
  });

  test("should contain prefix/suffix Adornment when props is passed", () => {
    const { container } = render(
      <Input
        value="Value"
        type="text"
        onChange={() => null}
        prefixAdornment={"https://"}
        suffixAdornment={".com"}
      />
    );
    expect(
      container.getElementsByClassName("input__prefixAdornment").length
    ).toBe(1);
    expect(
      container.getElementsByClassName("input__suffixAdornment").length
    ).toBe(1);
  });
});
