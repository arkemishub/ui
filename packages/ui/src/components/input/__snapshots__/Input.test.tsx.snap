// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex"
    >
      <div
        class="z-10 w-full px-2 flex items-center min-h-[40px] min-w-[120px] border border-neutral rounded-theme bg-background text-background-contrast focus-within:outline-offset-0 focus-within:border focus-within:border-primary focus-within:outline focus-within:outline-2 focus-within:outline-primary/30 focus-within:disabled:cursor-not-allowed input__container"
        data-testid="arke-container"
      >
        <input
          class="bg-transparent min-h-[40px] min-w-[120px] w-full px-2 outline-0 border-0 input"
          data-testid="arke-input"
          type="text"
          value="Value"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
